### PR TITLE
Fix conflicts with markdown plugins

### PR DESCRIPTION
**Pull request by mkaz**
There is a conflict with wpcom-markdown which is used in Jetpack  and other plugins which add markdown support.

Here's the test case that duplicates the issue:

1. Install Jetpack (or other wpcom markdown implementation) and enable markdown
2. Create post in Gutenberg with a couple paragraphs.
3. Open post in classic editor and click update
4. One big paragraph.

## Description

This change adds the filter on `wp_insert_post_data` since
edits in classic editor do not go through the rest API.

Additionally, it changes the call to remove markdown which
is more inclusive of all the actions and filters this markdown
implementation touches.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code has proper inline documentation.

### CODE REVIEW

File: lib/plugin-compat.php [hunk 1]
```diff
@@ -25,10 +25,12 @@
...
+	if ( gutenberg_content_has_blocks( implode( "\n", $post_data ) ) ) {
+        if ( class_exists ( 'WPCom_Markdown' ) ) {
+            WPCom_Markdown::get_instance()->unload_markdown_for_posts();
+        }
 	}
-	return $post;
+	return $post_data;
 }
-add_filter( 'rest_pre_insert_post', 'gutenberg_remove_wpcom_markdown_support' );
+add_filter( 'wp_insert_post_data', 'gutenberg_remove_wpcom_markdown_support', 1, 1 );
```
Comment (aduth) [inline lib/plugin-compat.php]: Needs to be so early in priority? Can't it just be one step before where Jetpack it filtering it at?

Comment (aduth) [reply lib/plugin-compat.php]: The fourth argument is `1` by default and can be omitted.

File: lib/plugin-compat.php [hunk 2]
```diff
@@ -25,10 +25,12 @@
  * @param  array $post      Post object which contains content to check for block.
  * @return array $post      Post object.
  */
-function gutenberg_remove_wpcom_markdown_support( $post ) {
-	if ( gutenberg_content_has_blocks( $post->post_content ) ) {
-		remove_post_type_support( 'post', 'wpcom-markdown' );
+function gutenberg_remove_wpcom_markdown_support( $post_data ) {
+	if ( gutenberg_content_has_blocks( implode( "\n", $post_data ) ) ) {
+        if ( class_exists ( 'WPCom_Markdown' ) ) {
```
Comment (aduth) [inline lib/plugin-compat.php]: No need for nesting `if`, just combine into the top-level condition.

Comment (aduth) [inline lib/plugin-compat.php]: Whitespace issues here (needs tabs, not spaces).

File: lib/plugin-compat.php [hunk 3]
```diff
@@ -25,10 +25,12 @@
  * @param  array $post      Post object which contains content to check for block.
  * @return array $post      Post object.
  */
-function gutenberg_remove_wpcom_markdown_support( $post ) {
-	if ( gutenberg_content_has_blocks( $post->post_content ) ) {
-		remove_post_type_support( 'post', 'wpcom-markdown' );
+function gutenberg_remove_wpcom_markdown_support( $post_data ) {
+	if ( gutenberg_content_has_blocks( implode( "\n", $post_data ) ) ) {
```
Comment (aduth) [inline lib/plugin-compat.php]: What is the shape of `$post_data`? From what I can tell, it's a compacted associative array of all properties:

https://github.com/WordPress/wordpress-develop/blob/fde455a89f9eaaae9513ca722f2de20420ca4bdd/src/wp-includes/post.php#L3487

Why do we test anything other than content?

**Review by aduth — CHANGES_REQUESTED** [review]

File: lib/plugin-compat.php (hunk 1 above)
Comment (mkaz) [reply lib/plugin-compat.php]: Updated.

File: lib/plugin-compat.php (hunk 2 above)
Comment (mkaz) [reply lib/plugin-compat.php]: Fixed.

Comment (mkaz) [reply lib/plugin-compat.php]: Fixed, wrong editor.

File: lib/plugin-compat.php (hunk 3 above)
Comment (mkaz) [reply lib/plugin-compat.php]: Fixed. I misread documentation that post_data was just the post_content in an array.  I'm not sure what I did on my initial inspection of the data to convince me that was the case.

**Comment by mkaz** [comment]
Thanks for the review @aduth - I rushed through the initial implementation and hadn't circled back yet to clean up. Updated with all the suggestions and fixed lint errors.

**Review by aduth — COMMENTED** [review]
I can confirm that this resolves the issue with paragraphs being stripped when saving in the classic editor.

I'm seeing that the classic editor inserts many paragraphs of its own in-between the blocks, which persist into the published copy. Seeing whether this is specific to Markdown module...

**Review by aduth — APPROVED** [review]
Extra paragraph tags are unrelated to Markdown module, tracked separately at #3901

**Pull request merged**